File: android/src/main/java/io/boxo/expo/ExpoBoxoSdkModule.kt
package io.boxo.expo

import android.os.Handler
import android.os.Looper
import io.boxo.data.models.MiniappData
import io.boxo.data.models.PageAnimation
import io.boxo.js.params.CustomEvent
import io.boxo.js.params.PaymentData
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import io.boxo.sdk.*

class ExpoBoxoSdkModule : Module() {
    private var handler: Handler? = null

    override fun definition() = ModuleDefinition {
        Name("ExpoBoxoSdk")

        handler = Handler(Looper.getMainLooper())

        Events("customEvent", "paymentEvent", "miniappLifecycle", "onAuth", "miniappList")

        Function("setConfig") { options: ConfigOptions ->
            val globalTheme: Config.Theme = when (options.theme) {
                "light" -> Config.Theme.LIGHT
                "dark" -> Config.Theme.DARK
                else -> Config.Theme.SYSTEM
            }
            Boxo.init(appContext.activityProvider!!.currentActivity.application)
                .setConfig(
                    Config.Builder()
                        .setClientId(options.clientId)
                        .setUserId(options.userId)
                        .sandboxMode(options.sandboxMode)
                        .multitaskMode(options.multitaskMode)
                        .setTheme(globalTheme)
                        .setLanguage(options.language)
                        .permissionsPage(options.showPermissionsPage)
                        .showClearCache(options.showClearCache)
                        .showAboutPage(options.showAboutPage)
                        .debug(options.isDebug)
                        .build()
                )
        }

        Function("openMiniapp") { options: MiniappOptions ->
            val miniapp: Miniapp = Boxo.getMiniapp(options.appId)
                .setCustomEventListener { _, miniapp, customEvent ->
                    sendEvent(
                        "customEvent", mapOf(
                            "appId" to miniapp.appId,
                            "requestId" to customEvent.requestId,
                            "type" to customEvent.type,
                            "errorType" to customEvent.errorType,
                            "payload" to customEvent.payload,
                        )
                    )
                }
                .setPaymentEventListener { _, miniapp, paymentData ->
                    sendEvent(
                        "paymentEvent", mapOf(
                            "appId" to miniapp.appId,
                            "transactionToken" to paymentData.transactionToken,
                            "miniappOrderId" to paymentData.miniappOrderId,
                            "amount" to paymentData.amount,
                            "currency" to paymentData.currency,
                            "extraParams" to paymentData.extraParams,
                            "hostappOrderId" to paymentData.hostappOrderId,
                            "status" to paymentData.status,
                        )
                    )
                }
                .setAuthListener { _, miniapp ->
                    sendEvent("onAuth", mapOf("appId" to miniapp.appId))
                }
                .setLifecycleListener(object : Miniapp.LifecycleListener {
                    override fun onClose(miniapp: Miniapp) {
                        sendEvent(
                            "miniappLifecycle", mapOf(
                                "appId" to miniapp.appId,
                                "lifecycle" to "onClose"
                            )
                        )
                    }

                    override fun onError(miniapp: Miniapp, message: String) {
                        sendEvent(
                            "miniappLifecycle", mapOf(
                                "appId" to miniapp.appId,
                                "lifecycle" to "onError",
                                "error" to message
                            )
                        )
                    }

                    override fun onLaunch(miniapp: Miniapp) {
                        sendEvent(
                            "miniappLifecycle", mapOf(
                                "appId" to miniapp.appId,
                                "lifecycle" to "onLaunch"
                            )
                        )
                    }

                    override fun onPause(miniapp: Miniapp) {
                        sendEvent(
                            "miniappLifecycle", mapOf(
                                "appId" to miniapp.appId,
                                "lifecycle" to "onPause"
                            )
                        )
                    }

                    override fun onResume(miniapp: Miniapp) {
                        sendEvent(
                            "miniappLifecycle", mapOf(
                                "appId" to miniapp.appId,
                                "lifecycle" to "onResume"
                            )
                        )
                    }

                    override fun onUserInteraction(miniapp: Miniapp) {
                        sendEvent(
                            "miniappLifecycle", mapOf(
                                "appId" to miniapp.appId,
                                "lifecycle" to "onUserInteraction"
                            )
                        )
                    }
                })
            if (options.data != null) miniapp.setData(options.data)
            val configBuilder = MiniappConfig.Builder()
            if (options.theme != null) {
                val miniappTheme: Config.Theme? = when (options.theme) {
                    "light" -> Config.Theme.LIGHT
                    "dark" -> Config.Theme.DARK
                    "system" -> Config.Theme.SYSTEM
                    else -> null
                }
                if (miniappTheme != null) {
                    configBuilder.setTheme(miniappTheme)
                }
            }
            options.extraUrlParams?.also { extraUrlParams ->
                val stringMap: MutableMap<String, String> = HashMap()
                for ((key, value) in extraUrlParams) stringMap[key] = value.toString()
                configBuilder.setExtraUrlParams(stringMap)
            }
            options.urlSuffix?.also { suffix -> configBuilder.setUrlSuffix(suffix) }
            options.colors?.also { colors ->
                configBuilder.setColors(
                    colors["primaryColor"] ?: "",
                    colors["secondaryColor"] ?: "",
                    colors["tertiaryColor"] ?: "",
                )
            }
            val pageAnimation = runCatching { PageAnimation.valueOf(options.pageAnimation ?: "") }
                    .getOrDefault(PageAnimation.BOTTOM_TO_TOP)
            configBuilder.pageAnimation(pageAnimation)
            configBuilder.enableSplash(options.enableSplash)
            configBuilder.saveState(options.saveState)
            miniapp.setConfig(configBuilder.build())
            miniapp.open(appContext.currentActivity!!)
        }

        Function("setAuthCode") { appId: String, authCode: String ->
            Boxo.getMiniapp(appId)
                .setAuthCode(authCode)
        }

        Function("setAuthTokens") { appId: String, tokens: Map<String, String> ->
            Boxo.getMiniapp(appId)
                .setAuthTokens(tokens)
        }

        Function("sendCustomEvent") { customEvent: CustomEventData ->
            handler?.post {
                Boxo.getMiniapp(customEvent.appId)
                    .sendEvent(
                        CustomEvent(
                            requestId = customEvent.requestId,
                            type = customEvent.type ?: "",
                            errorType = customEvent.errorType,
                            payload = customEvent.payload ?: emptyMap()
                        )
                    )
            }
        }

        Function("sendPaymentEvent") { paymentEvent: PaymentEventData ->
            handler?.post {
                Boxo.getMiniapp(paymentEvent.appId)
                    .sendPaymentResult(
                        PaymentData(
                            transactionToken = paymentEvent.transactionToken ?: "",
                            miniappOrderId = paymentEvent.miniappOrderId ?: "",
                            amount = paymentEvent.amount ?: 0.0,
                            currency = paymentEvent.currency ?: "",
                            status = paymentEvent.status ?: "",
                            hostappOrderId = paymentEvent.hostappOrderId ?: "",
                            extraParams = paymentEvent.extraParams
                        )
                    )
            }
        }

        Function("closeMiniapp") { appId: String ->
            Boxo.getExistingMiniapp(appId)?.close()
        }

        Function("hideMiniapps") {
            Boxo.hideMiniapps()
        }

        Function("logout") {
            Boxo.logout()
        }

        Function("getMiniapps") {
            Boxo.getMiniapps(object : MiniappListCallback {
                override fun onFailure(e: Exception) {
                    sendEvent("miniappList", mapOf("error" to e.toString()))
                }

                override fun onSuccess(miniapps: List<MiniappData>) {
                    sendEvent("miniappList", mapOf("miniapps" to miniapps.map {
                        mapOf(
                            "name" to it.name,
                            "description" to it.description,
                            "appId" to it.appId,
                            "logo" to it.logo,
                            "category" to it.category,
                        )
                    }))
                }
            })
        }
    }
}

File: ios/ExpoBoxoSdkModule.swift
import ExpoModulesCore
import BoxoSDK

public class ExpoBoxoSdkModule: Module {
    
    public func definition() -> ModuleDefinition {
        Name("ExpoBoxoSdk")
        
        Events("customEvent", "paymentEvent", "miniappLifecycle", "onAuth", "miniappList")
        
        Function("setConfig") { (options : ConfigOptions) in
            var globalTheme : Theme = .System
            switch (options.theme) {
            case "dark":
                globalTheme = .Dark
            case "light":
                globalTheme = .Light
            default:
                globalTheme = .System
            }
            
            let config = Config(clientId: options.clientId, theme: globalTheme)
            config.setUserId(id: options.userId)
            config.language = options.language
            config.sandboxMode = options.sandboxMode
            config.permissionsPage = options.showPermissionsPage
            config.showClearCache = options.showClearCache
            config.showAboutPage = options.showAboutPage
            
            Boxo.shared.setConfig(config: config)
        }
        
        Function("openMiniapp") { (options : MiniappOptions) in
            let miniApp = Boxo.shared.getMiniapp(appId: options.appId)
            miniApp.setData(data: options.data)
            miniApp.delegate = self
            
            let miniappConfig = MiniappConfig()
            miniappConfig.saveState = options.saveState
            miniappConfig.enableSplash(isSplashEnabled: options.enableSplash)
            miniappConfig.setExtraParams(extraParams: options.extraUrlParams)
            miniappConfig.urlSuffix = options.urlSuffix ?? ""
            
            switch options.pageAnimation {
                case "BOTTOM_TO_TOP":
                    miniappConfig.pageAnimation = .BOTTOM_TO_TOP
                case "TOP_TO_BOTTOM":
                    miniappConfig.pageAnimation = .TOP_TO_BOTTOM
                case "LEFT_TO_RIGHT":
                    miniappConfig.pageAnimation = .LEFT_TO_RIGHT
                case "RIGHT_TO_LEFT":
                    miniappConfig.pageAnimation = .RIGHT_TO_LEFT
                case "FADE_IN":
                    miniappConfig.pageAnimation = .FADE_IN
                default:
                    miniappConfig.pageAnimation = .BOTTOM_TO_TOP
            }

            if let theme = options.theme {
                switch theme {
                case "dark":
                    miniappConfig.setTheme(theme: .Dark)
                case "light":
                    miniappConfig.setTheme(theme: .Light)
                case "system":
                    miniappConfig.setTheme(theme: .System)
                default:
                    break
                }
            }
            
            if let colors = options.colors {
                miniappConfig.setColor(color: MiniappColor(primary: colors["primary_color"] ?? "", secondary: colors["secondary_color"] ?? "", tertiary: colors["tertiary_color"] ?? ""))
            }
            
            miniApp.setConfig(config: miniappConfig)
            DispatchQueue.main.async {
                miniApp.open(viewController: UIApplication.shared.delegate!.window!!.rootViewController!)
            }
        }
        
        Function("setAuthCode") { (appId : String, authCode: String) in
            DispatchQueue.main.async {
                Boxo.shared.getMiniapp(appId: appId).setAuthCode(authCode: authCode)
            }
        }

        Function("setAuthTokens") { (appId : String, tokens: [String : Any]) in
            DispatchQueue.main.async {
                Boxo.shared.getMiniapp(appId: appId).setAuthTokens(tokens)
            }
        }
        
        Function("sendCustomEvent") { (customEvent : CustomEventData) in
            let event = CustomEvent()
            event.requestId = customEvent.requestId ?? 0
            event.type = customEvent.type ?? ""
            event.errorType = customEvent.errorType ?? ""
            event.payload = customEvent.payload
            
            DispatchQueue.main.async {
                Boxo.shared.getMiniapp(appId: customEvent.appId).sendCustomEvent(customEvent: event)
            }
        }
        
        Function("sendPaymentEvent") { (paymentEvent : PaymentEventData) in
            let paymentData = PaymentData()
            paymentData.transactionToken = paymentEvent.transactionToken ?? ""
            paymentData.miniappOrderId = paymentEvent.miniappOrderId ?? ""
            paymentData.amount = paymentEvent.amount ?? 0.0
            paymentData.currency = paymentEvent.currency ?? ""
            paymentData.status = paymentEvent.status ?? ""
            paymentData.hostappOrderId = paymentEvent.hostappOrderId ?? ""
            paymentData.extraParams = paymentEvent.extraParams
            
            DispatchQueue.main.async {
                Boxo.shared.getMiniapp(appId: paymentEvent.appId).sendPaymentEvent(paymentData: paymentData)
            }
        }
        
        Function("closeMiniapp") { (appId : String) in
            if let miniapp = Boxo.shared.getExistingMiniapp(appId: appId) {
                DispatchQueue.main.async {
                    miniapp.close()
                }
            }
        }
        
        Function("hideMiniapps") {
            DispatchQueue.main.async {
                Boxo.shared.hideMiniapps()
            }
        }
        
        Function("logout") {
            DispatchQueue.main.async {
                Boxo.shared.logout()
            }
        }
        
        Function("getMiniapps") {
            Boxo.shared.getMiniapps { miniapps, error in
                if let error = error {
                    self.sendEvent("miniappList", ["error" : error])
                } else {
                    self.sendEvent("miniappList", ["miniapps" : miniapps.map { miniappData in
                        return [
                            "appId" : miniappData.appId,
                            "name" : miniappData.name,
                            "category" : miniappData.category,
                            "logo" : miniappData.logo,
                            "description" : miniappData.longDescription
                        ]
                    }])
                }
            }
        }
    }
}

extension ExpoBoxoSdkModule : MiniappDelegate {
    public func didReceivePaymentEvent(miniapp: Miniapp, paymentData: PaymentData) {
        let dict = [
            "appId" : miniapp.appId,
            "transactionToken" : paymentData.transactionToken,
            "miniappOrderId" : paymentData.miniappOrderId,
            "amount" : paymentData.amount,
            "currency" : paymentData.currency,
            "status" : paymentData.status,
            "hostappOrderId" : paymentData.hostappOrderId,
            "extraParams" : paymentData.extraParams ?? [String : Any]()
        ] as [String: Any]
        
        sendEvent("paymentEvent", dict)
    }
    
    public func didReceiveCustomEvent(miniapp: Miniapp, customEvent: CustomEvent) {
        let dict = [
            "appId" : miniapp.appId,
            "requestId" : customEvent.requestId,
            "type" : customEvent.type,
            "errorType" : customEvent.errorType,
            "payload" : customEvent.payload ?? [String : Any]()
        ] as [String: Any]
        
        sendEvent("customEvent", dict)
    }
    
    public func onLaunch(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onLaunch"
        ]
        
        sendEvent("miniappLifecycle", dict)
    }
    
    public func onResume(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onResume"
        ]
        
        sendEvent("miniappLifecycle", dict)
    }
    
    public func onPause(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onPause"
        ]
        
        sendEvent("miniappLifecycle", dict)
    }
    
    public func onClose(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onClose"
        ]
        
        sendEvent("miniappLifecycle", dict)
    }
    
    public func onError(miniapp: Miniapp, message: String) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onError",
            "error" : message
        ]
        
        sendEvent("miniappLifecycle", dict)
    }
    
    public func onUserInteraction(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onUserInteraction"
        ]
        
        sendEvent("miniappLifecycle", dict)
    }
    
    public func onAuth(miniapp: Miniapp) {
        let dict = [
            "appId" : miniapp.appId,
            "lifecycle" : "onAuth"
        ]
        
        sendEvent("onAuth", dict)
    }
}
